### ISSUE TITLE
Viewport Position after Resize

### ISSUE THREAD

**Issue by matt328** [glfw/glfw#2533]
I'm using glfw 3.4 with Vulkan 1.3.261 on Windows 10.  I've just switched from 'static' viewport and scissor specified in the Vulkan pipeline object to dynamic viewport and scissor specified when recording the command buffer, and now when I resize my window, this happens:

![image](https://github.com/glfw/glfw/assets/706207/8ea8c30e-2bae-4631-97ab-27b91524f345)

It looks like the position of the rendered area isn't taking into account the window decorations.  The strange part is when I move the window, even a single pixel, it corrects itself. Also strange is the ImGui controls' hitboxes are in the correct areas despite the rendered image being shifted.

One thing I suspect might be an issue is I'm only recreating my swapchain based on if acquireNextImage or present returns an error rather than by using the glfw framebuffersize callback. This worked before with static viewport and scissor, and it also works on Windows 11, so I'm not sure changing all that is worthwhile or not.

This project is a little beyond a small self contained example, so if there's any sections of code that would be helpful I can post or link to them.

**Comment by dougbinks** [comment]
I've labelled this as a support issue (question) as I think you're looking for help with Vulkan in achieving what you want.

If this is instead an issue report let me know, however GLFW itself doesn't interact with Vulkan other than to provide some assistance functions for cross platform device and surface creation.

You mention using ImGui: if you are using Dear ImGui perhaps try their GLFW + Vulkan example to check if it has the same issue, and also make sure to run with the debug validation layer.

**Comment by matt328** [comment]
Yeah, I wasn't really sure where to start with this. Was kinda hoping someone might have some insight into what dragging the window does in glfw that might cause this to correct itself. So yeah it's definitely a support question for now.

**Comment by dougbinks** [comment]
> when I resize my window, this happens

> Was kinda hoping someone might have some insight into what dragging the window does in glfw

When you resize a GLFW window the operating system / windowing system handles the resizing and GLFW simply receives event information. Everything else with Vulkan is up to the client code. I'm not sure what would cause the symptoms you see but would start with debug validation layers and comparing the difference between your application and a known working example (such as the Dear ImGui GLFW + Vulkan demo).

**Comment by dougbinks** [comment]
I was able to get time to test my own projects and also Dear ImGui GLFW + Vulkan demo (which my Vulkan code is somewhat based off), and both work as expected under Windows 11 with GLFW 3.4 and Vulkan SDK 1.3.275.

**Comment by JuanDiegoMontoya** [comment]
I'm experiencing this exact issue as well, in both [a Vulkan app](https://github.com/JuanDiegoMontoya/Frogfood/tree/hdr) and [an OpenGL one](https://github.com/JuanDiegoMontoya/Fwog/blob/main/example/06_msaa.cpp). The code I have for resizing in the OpenGL app is [pretty simple](https://github.com/JuanDiegoMontoya/Fwog/blob/main/example/common/Application.cpp#L129-L144), while the Vulkan one is [a bit more complicated](https://github.com/JuanDiegoMontoya/Frogfood/blob/hdr/src/Application.cpp#L702-L747).

For reference, I am on Window 10 using an RX 7900 XTX (in case this is a driver issue).

OpenGL application output after resizing the window slightly (the right and bottom of the framebuffer are corrupted):
![image](https://github.com/user-attachments/assets/c0571627-cca1-4fe1-9f30-85708545b74b)

Vulkan application output displaying similar artifacts:
![image](https://github.com/user-attachments/assets/863055bd-1f0c-469d-a70c-95afb5729d1e)

**Comment by dougbinks** [comment]
@JuanDiegoMontoya Do any of the GLFW tests show the same issue, and have you tried testing the  Dear ImGui GLFW + Vulkan demo?

I would also check that have your viewport set up correctly.

**Comment by deccer** [comment]
Oddly enough I can resize @JuanDiegoMontoya's OpenGL and FrogRender apps just fine, Win11 (22631) + RTX4070 (537.42 (not latest))

**Comment by JuanDiegoMontoya** [comment]
@dougbinks I get the same issue in the "triangle-vulkan" test for GLFW. Attached is a video of the issue in action.
![image](https://github.com/user-attachments/assets/4345319e-291c-4904-9d67-d5d1c6dd6169)

[triangle-vulkan_o1ZsHyUEZ0.zip](https://github.com/user-attachments/files/16597592/triangle-vulkan_o1ZsHyUEZ0.zip)

**Comment by dougbinks** [comment]
On my machine (Win11 + 7900 XTX with 24.5.1 driver) the `triangle-vulkan` test resizes properly, although during resizing there is a slight lag between the window size and the rendered viewport.

The resizing is triggered by `WM_SIZE` message event in windows, which is ancient and should work with Windows 10 as well as Windows 11 (indeed before I upgraded it worked for me).

I'll have a think, but first I'll upgrade to the same driver as you to test.

Could you test the OpenGL `boing` example (no video required)? This should resize properly after the window has stopped resizing.

**Comment by JuanDiegoMontoya** [comment]
Same issue in the boing example.

![image](https://github.com/user-attachments/assets/02898032-0ca2-4f07-af1c-2d37d18a8a2b)

**Comment by dougbinks** [comment]
Can I just check that this is after resizing, not during (no video or image required)?

**Comment by JuanDiegoMontoya** [comment]
This is indeed after resizing. The behavior of every app I tested is the same as the original one by @matt328:
- Resizing the window any amount makes it seemingly "forget" about the decorations, leading to the pictured artifacts
- Moving the window any amount, maximizing + restoring down, or minimizing and un-minimizing the window "fixes" the issue (until it's resized again)
- A janky workaround that I found to "work" is to call `glfwSetWindowPos` twice, like so:
```cpp
  int x, y;
  glfwGetWindowPos(window, &x, &y);
  glfwSetWindowPos(window, x + 1, y);
  glfwSetWindowPos(window, x, y);
```
but the window still exhibits flickering until the resizing is complete (the user releases left click).

**Comment by deccer** [comment]
I'm getting a white region inside the newly extended window too the first time i resize the window, any resize afterwards (slowly or erratically) shows the normal black background in the newly resized areas before the renderer refreshes the whole size after releasing the mouse

**Comment by dougbinks** [comment]
After updating to AMD driver 24.7.1 I still cannot replicate the issue.

I am wondering if there might be a [game hook such as Nahimic](https://github.com/ocornut/imgui/issues/4542), which is known to cause issues during resizing? If you put a breakpoint in the code and then when paused look at modules you can see any hooks. It can be useful to sort by Path in order to see the non Windows system ones.

**Comment by dougbinks** [comment]
> Resizing the window any amount makes it seemingly "forget" about the decorations, leading to the pictured artifacts

This leads me to believe this is likely outside the scope of GLFW, and a driver or other issue since GLFW cannot draw to the window decorations.

**Comment by dougbinks** [comment]
@deccer:

> I'm getting a white region inside the newly extended window too the first time i resize the window, any resize afterwards (slowly or erratically) shows the normal black background in the newly resized areas before the renderer refreshes the whole size after releasing the mouse

So long as the application renders correctly after resizing is complete this is the expected behaviour (handling resizing during resizing on windows is complicated).

**Comment by deccer** [comment]
Yep, it does. I too believe this is normal behavior. Resizing is indeed a rabbit hole :)

**Comment by JuanDiegoMontoya** [comment]
> After updating to AMD driver 24.7.1 I still cannot replicate the issue.
> 
> I am wondering if there might be a [game hook such as Nahimic](https://github.com/ocornut/imgui/issues/4542), which is known to cause issues during resizing? If you put a breakpoint in the code and then when paused look at modules you can see any hooks. It can be useful to sort by Path in order to see the non Windows system ones.

I don't see any modules outside System32 except for my own app and LLVM-C.dll. [modules.txt](https://github.com/user-attachments/files/16598185/new.1.txt)


> > Resizing the window any amount makes it seemingly "forget" about the decorations, leading to the pictured artifacts
> 
> This leads me to believe this is likely outside the scope of GLFW, and a driver or other issue since GLFW cannot draw to the window decorations.

The decorations aren't being drawn to, rather everything is be shifted up and left by an amount that *seems* to match the size of the decorations.

**Comment by JuanDiegoMontoya** [comment]
With your hint, I managed to solve the issue @dougbinks!

However, I'm not sure what fixed it exactly since undoing the steps I took does not reproduce the issue again. This is what I did:

- In task manager, disable the following apps from running on startup:
  - AMD Noise Suppression
  - ccProfilerTray.exe
  - CalibrationLoader 5.0 Application
- Reboot PC (I tried rebooting alone earlier and it did not fix the issue)
- Running Visual Studio 2022 and launching the glfw `boing` example.

I did make an interesting observation after the fact though. When my monitor has HDR enabled in Windows (which was the case this whole time), resizing `boing` behaves ideally (as though you were taking a texture and scaling it with bilinear filtering). In SDR mode, there remains a little jank as black or white edges appear in the new area (when expanding the window size) or the edges are cropped (when reducing the size). I expect a little jank when resizing, so this is acceptable.

I'm sorry I can't say what was causing the issue precisely, but I guess it was one of the listed startup programs behaving poorly at the time.

**Comment by dougbinks** [comment]
Glad you were able to solve this!
